probe-rs/src/architecture/arm/core/armv7a.rs

```diff
@@ -297,29 +274,271 @@ impl<'probe> Armv7a<'probe> {
+
+    Ok(())
+}
 
-        Ok(dbgdscr.halted())
+/// If a specified breakpoint is set, disable it and clear it.
+pub(crate) fn clear_hw_breakpoint(
+    memory: &mut dyn ArmMemoryInterface,
+    base_address: u64,
+    bp_unit_index: usize,
+) -> Result<(), ArmError> {
+    let bp_value_addr = Dbgbvr::get_mmio_address_from_base(base_address)?
+        + (bp_unit_index * size_of::<u32>()) as u64;
+    let bp_control_addr = Dbgbcr::get_mmio_address_from_base(base_address)?
+        + (bp_unit_index * size_of::<u32>()) as u64;
+
+    memory.write_word_32(bp_value_addr, 0)?;
+    memory.write_word_32(bp_control_addr, 0)?;
+    Ok(())
+}
+
+/// Get a specific hardware breakpoint. If the breakpoint is not set, return `None`.
+pub(crate) fn get_hw_breakpoint(
+    memory: &mut dyn ArmMemoryInterface,
+    base_address: u64,
+    bp_unit_index: usize,
+) -> Result<Option<u32>, ArmError> {
+    let bp_value_addr = Dbgbvr::get_mmio_address_from_base(base_address)?
+        + (bp_unit_index * size_of::<u32>()) as u64;
+    let bp_value = memory.read_word_32(bp_value_addr)?;
+
+    let bp_control_addr = Dbgbcr::get_mmio_address_from_base(base_address)?
+        + (bp_unit_index * size_of::<u32>()) as u64;
+    let bp_control = Dbgbcr(memory.read_word_32(bp_control_addr)?);
+
+    Ok(if bp_control.e() { Some(bp_value) } else { None })
+}
+
+/// Execute a single instruction.
+fn execute_instruction(
+    memory: &mut dyn ArmMemoryInterface,
+    base_address: u64,
+    instruction: u32,
+) -> Result<Dbgdscr, ArmError> {
+    // Run instruction
+    let address = Dbgitr::get_mmio_address_from_base(base_address)?;
+    memory.write_word_32(address, instruction)?;
+
+    // Wait for completion
+    let address = Dbgdscr::get_mmio_address_from_base(base_address)?;
+    let mut dbgdscr = Dbgdscr(memory.read_word_32(address)?);
+
+    while !dbgdscr.instrcoml_l() {
+        dbgdscr = Dbgdscr(memory.read_word_32(address)?);
+    }
+
+    // Check if we had any aborts, if so clear them and fail
+    if dbgdscr.adabort_l() || dbgdscr.sdabort_l() {
+        let address = Dbgdrcr::get_mmio_address_from_base(base_address)?;
+        let mut dbgdrcr = Dbgdrcr(0);
+        dbgdrcr.set_cse(true);
+
+        memory.write_word_32(address, dbgdrcr.into())?;
+        return Err(ArmError::Armv7a(
+            crate::architecture::arm::armv7a::Armv7aError::DataAbort,
+        ));
+    }
+
+    Ok(dbgdscr)
+}
+
+/// Set the DBGDBGDTRRX register, which can be accessed with an
+/// `STC p14, c5, ..., #4` instruction.
+fn set_instruction_input(
+    memory: &mut dyn ArmMemoryInterface,
+    base_address: u64,
+    value: u32,
+) -> Result<(), ArmError> {
+    // Move value
+    let address = Dbgdtrrx::get_mmio_address_from_base(base_address)?;
+    memory.write_word_32(address, value)?;
+
+    // Wait for RXfull
+    let address = Dbgdscr::get_mmio_address_from_base(base_address)?;
+    let mut dbgdscr = Dbgdscr(memory.read_word_32(address)?);
+
+    while !dbgdscr.rxfull_l() {
```

Comment:
Not related to this PR but eventually it would be nice if we could take advantage of CMSIS-DAP's value matching functionality, and do this while loop on the probe. It would need some MemoryInterface/CoreInterface changes I believe, but if someone wants to think about the problem, they'd be welcome to :)

My concern is: can this accidentally turn into an infinite loop? I'd like to see some timeout on unconstrained loops, please.